ssl/quic/quic_ackm.c

```diff
@@ -0,0 +1,1997 @@
+                    && (z->prev == NULL || start > z->prev->range.end)) {
+            if (z->range.start == end + 1) {
+                /* We can extend the following range backwards. */
+                z->range.start = start;
+
+                /*
+                 * If this closes a gap we now need to merge
+                 * consecutive nodes.
+                 */
+                pn_set_merge_adjacent(s, z);
+            } else if (z->prev != NULL && z->prev->range.end + 1 == start) {
+                /* We can extend the preceding range forwards. */
+                z->prev->range.end = end;
+
+                /*
+                 * If this closes a gap we now need to merge
+                 * consecutive nodes.
+                 */
+                pn_set_merge_adjacent(s, z);
+            } else {
+                /*
+                 * The new interval is between intervals without overlapping or
+                 * touching them, so insert between, preserving sort.
+                 */
+                x = OPENSSL_zalloc(sizeof(struct pn_set_item_st));
+                if (x == NULL)
+                    return 0;
+
+                x->range.start = start;
+                x->range.end   = end;
+
+                x->next = z;
+                x->prev = z->prev;
+                if (x->prev != NULL)
+                    x->prev->next = x;
+                z->prev = x;
+                if (s->head == z)
+                    s->head = x;
+
+                ++s->num_ranges;
+            }
+            break;
+        }
+    }
+
+    return 1;
+}
+
+/*
+ * Remove a range from the set. Returns 0 on allocation failure, in which case
+ * the PN set is unchanged. Otherwise, returns 1. Ranges which are not already
+ * in the set can be removed without issue. If a passed range is not in the PN
+ * set at all, this is a no-op and returns 1.
+ */
+static int pn_set_remove(struct pn_set_st *s, const OSSL_QUIC_ACK_RANGE *range)
+{
+    struct pn_set_item_st *z, *zprev, *y;
+    QUIC_PN start = range->start, end = range->end;
+
+    if (!ossl_assert(start <= end))
+        return 0;
+
+    /* Walk backwards since we will most often be removing at the end. */
+    for (z = s->tail; z != NULL; z = zprev) {
+        zprev = z->prev;
+
+        if (z->range.end < start)
+            /* No overlapping ranges can exist beyond this point, so stop. */
+            break;
+
+        if (start <= z->range.start && end >= z->range.end) {
+            /*
+             * The range being removed dwarfs this range, so it should be
+             * removed.
+             */
+            if (z->next != NULL)
+                z->next->prev = z->prev;
+            if (z->prev != NULL)
+                z->prev->next = z->next;
+            if (s->head == z)
+                s->head = z->next;
+            if (s->tail == z)
+                s->tail = z->prev;
+
+            OPENSSL_free(z);
+            --s->num_ranges;
+        } else if (z->range.start >= start) {
```

Comment:
It would make it easier to reason about the comparisons in this set of "ifs" if they were in the same order, e.g. on line 689 above we compare `start <= r->range.start`, but here we do `z->range.start >= start`. Similarly we have `end >= z->range.end` above, and `z->range.end <= end` below.